File: Trees/236. 二叉树的最近公共祖先/236. 二叉树的最近公共祖先.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        # 終止條件
        if not root or root == p or root == q : return root

        L = self.lowestCommonAncestor(root.left, p, q)
        R = self.lowestCommonAncestor(root.right, p, q)

        # L、R皆存在
        if L and R : return root
        # 僅L存在，R為空
        elif not R : return L
        # 僅R存在，L為空
        else : return R

File: Trees/236. 二叉树的最近公共祖先/236. 二叉树的最近公共祖先.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        // 終止條件
        if(root == NULL || root == p ||root == q) return root;

        TreeNode* L = lowestCommonAncestor(root->left, p, q);
        TreeNode* R = lowestCommonAncestor(root->right, p, q);

        // 都不為NULL
        if(L != NULL && R != NULL) return root;
        // L為NULL
        else if(L == NULL && R != NULL) return R;
        // R為NULL
        else return L;
    }
};